ddtrace/internal/codeowners.py

```diff
@@ -0,0 +1,196 @@
+
+    regex += "/" if matches_dir else r"(?:\Z|/)"
+    return re.compile(regex)
+
+
+class Codeowners(object):
+    """Provide interface to parse CODEOWNERS file and match a given path against it."""
+
+    KNOWN_LOCATIONS = (
+        "CODEOWNERS",
+        ".github/CODEOWNERS",
+        "docs/CODEOWNERS",
+        ".gitlab/CODEOWNERS",
+    )
+
+    def __init__(self, path=None, cwd=None):
+        # type: (Optional[str], Optional[str]) -> None
+        """Initialize Codeowners object.
+
+        :param path: path to CODEOWNERS file otherwise try to use any from known locations
+        """
+        path = path or self.location(cwd)
+        if path is not None:
+            self.path = path  # type: str
+        self.patterns = []  # type: List[Tuple[re.Pattern, List[str]]]
+        self.parse()
+
+    def location(self, cwd=None):
+        # type: (Optional[str]) -> Optional[str]
+        """Return the location of the CODEOWNERS file."""
+        cwd = cwd or os.getcwd()
+        for location in self.KNOWN_LOCATIONS:
+            path = os.path.join(cwd, location)
+            if os.path.isfile(path):
+                return path
+        raise ValueError("CODEOWNERS file not found")
+
+    def parse(self):
+        # type: () -> None
+        """Parse CODEOWNERS file and store the lines and regexes."""
+        with open(self.path) as f:
+            patterns = []
+            for line in f.readlines():
+                line = line.strip()
+                if line == "":
+                    continue
+                # Lines starting with '#' are comments.
+                if line.startswith("#"):
+                    continue
+                if line.startswith("[") and line.endswith("]"):
+                    # found a code owners section
+                    continue
+                if line.startswith("^[") and line.endswith("]"):
+                    # found an optional code owners section
+                    continue
+
+                elements = iter(line.split())
+                path = next(elements, None)
+                if path is None:
+                    continue
+
+                try:
+                    pattern = path_to_regex(path)
+                except ValueError:
+                    continue
+
+                owners = []
+                for owner in elements:
+                    if owner:
+                        owners.append(owner)
+
+                if not owners:
+                    continue
+                patterns.append((pattern, owners))
+            # Order is important. The last matching pattern has the most precedence.
+            patterns.reverse()
+            self.patterns = patterns
+
+    def of(self, path):
+        # type: (str) -> List[str]
+        """Return code owners for a given path.
+
+        :param path: path to check
+        :return: list of file code owners identified by the given path
+        """
+        for pattern, owners in self.patterns:
```

Comment:
is it worth caching the result of this method? do we expect these lookups to be repeated frequently for the same path?

Reply:
I don't think it is necessary at the moment. Each integration can have its caching strategy if needed.